### PR TITLE
#178: _loadMask and _loadFrame function events

### PR DESCRIPTION
**Pull request by BeeMargarida**
| - | - |
| --- | --- |
| Issue | https://github.com/ripe-tech/ripe-white/issues/178 |
| Decisions | Added `pre_frame` and `post_frame`events. |
| Animated GIF | - |

### CODE REVIEW

File: src/js/visual/configurator.js [hunk 1]
```diff
@@ -972,17 +978,22 @@ ripe.Configurator.prototype._loadMask = function(maskImage, view, position, opti
...
             ? () => {
+                  this.trigger("post_mask", maskImage, view, position);
                   setTimeout(() => {
                       this._drawMask(maskImage);
                   }, 150);
               }
-            : null;
+            : () => {
+                  this.trigger("post_mask", maskImage, view, position);
+                  return null;
```
Comment (gcandal) [inline src/js/visual/configurator.js:995]: No need, you can just turn this into a 1-liner

File: src/js/visual/configurator.js [hunk 2]
```diff
@@ -912,6 +914,7 @@ ripe.Configurator.prototype._loadFrame = async function(view, position, options
         if (isReady) {
             await this._drawFrame(image, animate, duration);
         }
+        this.trigger("post_frame", view, position, options);
```
Comment (gcandal) [inline src/js/visual/configurator.js:917]: Maybe we should include a boolean in the arguments saying if this was a no-op (the same applies to the mask when it doesn't actually load a new image).

File: src/js/visual/configurator.js [hunk 3]
```diff
@@ -948,12 +951,15 @@ ripe.Configurator.prototype._loadFrame = async function(view, position, options
...
     await imagePromise;
+
+    this.trigger("post_frame", view, position, true, options);
 };
 
 /**
  * @ignore
  */
 ripe.Configurator.prototype._loadMask = function(maskImage, view, position, options) {
+    this.trigger("pre_mask", maskImage, view, position);
```
Comment (gcandal) [inline src/js/visual/configurator.js:962]: Similarly to `pre_frame` this must also include `options`.

File: src/js/visual/configurator.js [hunk 4]
```diff
@@ -972,17 +978,19 @@ ripe.Configurator.prototype._loadMask = function(maskImage, view, position, opti
         color: backgroundColor
     });
     if (draw && maskImage.dataset.src === url) {
+        this.trigger("post_mask", maskImage, view, position, false);
```
Comment (gcandal) [inline src/js/visual/configurator.js:981]: Same as above

File: src/js/visual/configurator.js [hunk 5]
```diff
@@ -912,6 +914,7 @@ ripe.Configurator.prototype._loadFrame = async function(view, position, options
         if (isReady) {
             await this._drawFrame(image, animate, duration);
         }
+        this.trigger("post_frame", view, position, false, options);
```
Comment (gcandal) [inline src/js/visual/configurator.js:917]: Maybe move the `false` to the end so that we have (event name, operation args..., op/no-op).

**Comment by gcandal** [comment]
With https://github.com/ripe-tech/ripe-sdk/commit/5966310ef17c5eec4c17ce4cb8e5d1ec11ca874d this one is no longer needed, right @NFSS10 ?

**Comment by BeeMargarida** [comment]
> With [5966310](https://github.com/ripe-tech/ripe-sdk/commit/5966310ef17c5eec4c17ce4cb8e5d1ec11ca874d) this one is no longer needed, right @NFSS10 ?
@gcandal We think it is still needed since the `post_udpate` waits for all the updates from its children, given the total time. In this case, we want to catch the time each frame takes to update.

**Comment by gcandal** [comment]
@BeeMargarida ok, please remove the masks part and inline the value of `NO_OP` rather than using a variable then.

File: src/js/visual/configurator.js [hunk 6]
```diff
@@ -912,6 +914,7 @@ ripe.Configurator.prototype._loadFrame = async function(view, position, options
         if (isReady) {
             await this._drawFrame(image, animate, duration);
         }
+        this.trigger("post_frame", view, position, options, true);
```
Comment (gcandal) [inline src/js/visual/configurator.js:917]: Aren't these `true`/`false` the opposite of what they are in `post_update`?
- True: updated happened
- False: "cache" hit / no-op

File: src/python/ripe_demo/static/js/main.js [hunk 7]
```diff
@@ -4,7 +4,7 @@
  */
 var FACES = ["side", "top", "front"];
 
-window.onload = function() {
```
Comment (gcandal) [inline src/python/ripe_demo/static/js/main.js:7]: I think you need to `npm i`, you aren't using the latest dependencies that's why this happens. In any case, if something like this must be done in the future, please include it in a separate PR.

**Pull request merged**